src/cryptonote_protocol/quorumnet.cpp

```diff
@@ -0,0 +1,1392 @@
+            continue;
+        }
+
+        pinfo.relay_to_peers("vote", serialize_vote(vote));
+        votes_relayed++;
+    }
+    MDEBUG("Relayed " << votes_relayed << " votes");
+}
+
+void handle_vote(SNNetwork::message &m, void *self) {
+    auto &snw = *reinterpret_cast<SNNWrapper *>(self);
+
+    MDEBUG("Received a relayed vote from " << as_hex(m.pubkey));
+
+    if (m.data.size() != 1) {
+        MINFO("Ignoring vote: expected 1 data part, not " << m.data.size());
+        return;
+    }
+
+    try {
+        quorum_vote_t vote = deserialize_vote(m.data[0]);
+
+        if (vote.block_height > snw.core.get_current_blockchain_height()) {
+            MDEBUG("Ignoring vote: block height " << vote.block_height << " is too high");
+            return;
+        }
+
+        cryptonote::vote_verification_context vvc = {};
+        snw.core.add_service_node_vote(vote, vvc);
+        if (vvc.m_verification_failed)
+        {
+            MWARNING("Vote verification failed; ignoring vote");
+            return;
+        }
+
+        if (vvc.m_added_to_pool)
+            relay_votes(self, {{vote}});
+    }
+    catch (const std::exception &e) {
+        MWARNING("Deserialization of vote from " << as_hex(m.pubkey) << " failed: " << e.what());
+    }
+}
+
+/// Gets an integer value out of a bt_dict, if present and fits (i.e. get_int<> succeeds); if not
+/// present or conversion falls, returns `fallback`.
+template <typename I>
+std::enable_if_t<std::is_integral<I>::value, I> get_or(bt_dict &d, const std::string &key, I fallback) {
+    auto it = d.find(key);
+    if (it != d.end()) {
+        try { return get_int<I>(it->second); }
+        catch (...) {}
+    }
+    return fallback;
+}
+
+// Obtains the blink quorums, verifies that they are of an acceptable size, and verifies the given
+// input quorum checksum matches the computed checksum for the quorums (if provided), otherwise sets
+// the given output checksum (if provided) to the calculated value.  Throws std::runtime_error on
+// failure.
+quorum_array get_blink_quorums(uint64_t blink_height, const service_node_list &snl, const uint64_t *input_checksum, uint64_t *output_checksum = nullptr) {
+    // We currently just use two quorums, Q and Q' in the whitepaper, but this code is designed to
+    // work fine with more quorums (but don't use a single subquorum; that could only be secure or
+    // reliable but not both).
+    quorum_array result;
+
+    uint64_t local_checksum = 0;
+    for (uint8_t qi = 0; qi < NUM_BLINK_QUORUMS; qi++) {
+        auto height = blink_tx::quorum_height(blink_height, static_cast<blink_tx::subquorum>(qi));
+        if (!height)
+            throw std::runtime_error("too early in blockchain to create a quorum");
+        result[qi] = snl.get_quorum(quorum_type::blink, height);
+        auto &v = result[qi]->validators;
+        if (v.size() < BLINK_MIN_VOTES || v.size() > BLINK_SUBQUORUM_SIZE)
```

Comment:
BLINK_MIN_VOTES is 7, so it's not a huge burden, but indeed if this throws it is caught and returns an error to the sender.

Reply:
Does core_rpc_server wrap calls in try catch? It might not, I see individual try catch blocks in the code.

So in `on_send_raw_tx -> handle_blink_tx -> quorumnet_send_blink -> send_blink -> get_quorum_blinks` throwing

Reply:
I guess I can just test it

Reply:
Yeah calls in core_rpc_server don't handle exceptions unless explicitly caught.

Reply:
Will update tomorrow.

Reply:
Fixed; exceptions now get caught and propagated through the promise (which then get caught in core_rpc_server to return the exception message).